feat(NotificationMarker): Add `NotificationMarker` component

r100-stack opened the pull request:
Closes #348

* Adds `NotificationMarker` component.
* `NotificationMarker` adds a small notification circle to the top-right of the passed `children` prop. This can be applied to pretty much anything but is mostly intended for icons within default / borderless buttons.
* Prerequisite: PR #895 

## Sample Uses

#### 1. Primary Intended Use-case

[vokoscreenNG-2022-09-20_10-21-41.webm](https://user-images.githubusercontent.com/45748283/191283495-36c80f60-42c0-4533-8f4c-17f2ac1a3349.webm)

```html
<IconButton styleType='borderless'>
  <NotificationMarker><SvgNotification /></NotificationMarker>
</IconButton>
```

#### 2. Example with non-icon children

[vokoscreenNG-2022-09-20_10-22-45.webm](https://user-images.githubusercontent.com/45748283/191283744-2e937987-4550-4cd6-8e65-0e7bfa0d1acd.webm)

```html
<NotificationMarker type='positive' pulsing={true}>Live</NotificationMarker>
<NotificationMarker type='negative' pulsing={true}>Rec</NotificationMarker>
<NotificationMarker type='primary' pulsing={true}>1 new message</NotificationMarker>
```

#### 3. Using `enabled` prop

[vokoscreenNG-2022-09-20_10-28-24.webm](https://user-images.githubusercontent.com/45748283/191285303-062aabfe-ee93-4791-9ad4-a3bee618ae39.webm)

```jsx
let [demoNotificationCount, ...] = useState(0);
...
<NotificationMarker enabled={demoNotificationCount > 0}>
  <SvgNotification />
</NotificationMarker>
```

## Checklist

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] Approve test images for new stories
- [x] Add screenshots of the key elements of the component

r100-stack committed:
Ran yarn createComponent

r100-stack committed:
Working NotificationMarker example, but with workarounds. Need to make it work normally

r100-stack committed:
Found why NotificationMarker was disappearing when inside IconButton. Need to decide a logical way to implement it in prod

r100-stack committed:
Story with 4 NotificationMarkers

r100-stack committed:
Print props on click

r100-stack committed:
Added variant and urgent args

r100-stack committed:
Cleaned story

r100-stack committed:
Added a CustomChild story

r100-stack committed:
Added CommonProps to NotificationMarkerProps

r100-stack committed:
Removed duplicate className prop

r100-stack committed:
Added active prop. Also added JSDocs

r100-stack committed:
Renamed variant prop to type prop since Alert also has type

r100-stack committed:
Cleaned code for PR. Small corrections

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Added ProgrammaticDisplay story

r100-stack committed:
Added basic NotificationMarker test

r100-stack committed:
Added test to check props

r100-stack committed:
Added test to check inactive case

r100-stack committed:
Added test to support all kinds of children

r100-stack committed:
Added new stories to visual test

r100-stack committed:
Fixed type error. Improved ProgrammaticDisplay story

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Primary use case is first example in JSDocs

r100-stack committed:
Added test images

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,78 @@
+  } = props;
+  useTheme();
+  return (
+    <div
```
divs inside buttons is invalid markup. Does it still work with `<span>`?

rohan-kadkol replied:
To confirm, are you referring to what is mentioned in this [StackOverflow answer](https://stackoverflow.com/a/12982293/11547064)? This concept is new to me. I have a few questions:

* What is "markup" exactly?
* What is the meaning of "valid" markup?
* Why `<div>` inside `<button>` is invalid markup?
* How to know what is valid markup and what isn't? Do we always have to put it in a [validator tool](https://validator.w3.org/#validate_by_input+with_options) like the one mentioned in the answer? Or is there a website/source with rules for valid markup?

mayank99 replied:
markup is just html. "valid" means it's not allowed and will not be properly announced by assistive technologies (e.g. screen readers). and yes you need to validate it through a tool if you're not sure.
https://validator.w3.org/docs/help.html#validation_basics

there is an html spec but you're not expected to read the whole thing all at once 😄 https://html.spec.whatwg.org/

rohan-kadkol replied:
**[DONE]** Changed div to span and it seems to do the same effect.

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,78 @@
+  return (
+    <div
+      className={cx(
+        active
+          ? { [`iui-notification-${type}`]: true, 'iui-urgent': urgent }
+          : {},
```
```suggestion
          { [`iui-notification-${type}`]: active, 'iui-urgent': active && urgent }
```

rohan-kadkol replied:
**[DONE]** Nice idea, thanks

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,78 @@
+/**
+ * A small notification circle to the top-right of the passed children prop. This can be applied to pretty much anything but mostly intended for icons within default / borderless buttons.
+ * @example
+ * // Primary Intended Use-case
```
No need for this `//` comment

rohan-kadkol replied:
To confirm, did you mean to remove line 46? Or just remove the "`//`" from line 46?

I thought you meant the former, so I removed line 46. But just confirming if that's what you meant.

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,78 @@
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * A small notification circle to the top-right of the passed children prop. This can be applied to pretty much anything but mostly intended for icons within default / borderless buttons.
```
Line wrapping would help here

rohan-kadkol replied:
**[DONE]** Split line 44 into two lines

mayank99 commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,125 @@
+  },
+};
+
+export const ProgrammaticDisplay: Story<
```
I don't think this story is needed. Everything is "programmatic" if you think about it.

rohan-kadkol replied:
I removed this `ProgrammaticDisplay` story. Do you think there should be documentation, a story, or something similar to show how to use the `enabled` (currently called `active`) prop? Because without mentioning it, people might do something like:

```jsx
{numNotifications > 0 ? <NotificationMarker>child</NotificationMarker> : child}
```

instead of:

```jsx
{<NotificationMarker enabled={numNotifications > 0}>child</NotificationMarker>}
```

Using the first way would lead to an extra dom `<div>` when numNotifications > 0. This might cause CSS styling problems (users' direct descendent styles with `>`) and also maybe longer untidy code

elephantcatdog replied:
You have a good example on the property itself. Not sure how else to disseminate that info with the current documentation setup we have.

mayank99 replied:
I don't think it's necessarily a big deal if users do a ternary to toggle the notification-marker, so i think it's fine if we don't even show an example tbh

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,78 @@
+   * 'negative' = red,
+   * @default 'primary'
+   */
+  type?: 'primary' | 'positive' | 'warning' | 'negative';
```
I think this prop should be called `status`

rohan-kadkol replied:
**[DONE]** Ohh, I think I got confused with the `type` from `Alert.tsx`. Changed it to `status` now

![image](https://user-images.githubusercontent.com/45748283/191323870-7dee5a32-0777-4061-86d6-f6115a3ff8ed.png)

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,78 @@
+   *   <SvgNotification />
+   * </NotificationMarker>
+   */
+  active?: boolean;
```
`active` can be confused with selected state. Maybe `enabled` would be clearer

rohan-kadkol replied:
**[DONE]** Ohh, right, renamed it to `enabled`

mayank99 replied:
dont forget the jsdoc

rohan-kadkol replied:
**[DONE]** Oops, thanks, changed now

r100-stack committed:
Sorted index.ts in alphabetical order

rohan-kadkol commented on packages/iTwinUI-react/src/core/index.ts:
```diff
@@ -311,3 +311,6 @@ export {
 export type { ThemeType, MiddleTextTruncationProps } from './utils';
+
+export { NotificationMarker } from './NotificationMarker';
+export type { NotificationMarkerProps } from './NotificationMarker';
```
**[DONE]** Ohh, I forgot it, thanks

rohan-kadkol commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,125 @@
+  },
+};
+
+export const CustomChild: Story<NotificationMarkerProps> = (args) => {
```
I was trying to have a story to show that we can pass anything as a child instead of just SVGs like `<SvgNotification />`.

Maybe `CustomChild` is not an accurate name. What name do you recommend for such a story?

rohan-kadkol replied:
I was trying to have a story to show that we can pass anything as a child instead of just SVGs like `<SvgNotification />`.

Maybe `CustomChild` is not an accurate name. What name do you recommend for such a story?

elephantcatdog replied:
Maybe you could just put one text example into Basic? And replace the Custom example with a Pulsing example. Because in both Basic and Custom examples, the notification marker itself is basic.

rohan-kadkol replied:
**[DONE]** Ohh, nice idea, thanks. I changed `CustomChild` story to `Pulsing`.

`Basic` now shows one icon button and one text. `Pulsing` shows four texts (for each status) with pulsing notifications

rohan-kadkol replied:
**UPDATE**: I now created another story called `Status`. So now, `Basic` is same as before, `Pulsing` shows one icon and one bold text (similar to Basic), and `Status` shows all four statuses for icons and texts at once

r100-stack committed:
Renamed type to status

r100-stack committed:
Removed Programmatic Display story

r100-stack committed:
Cleaned JSDocs

r100-stack committed:
Renamed active to enabled

r100-stack committed:
Renamed type to status in one remaining place

r100-stack committed:
Changed child of NotificationMarker from div to span

r100-stack committed:
Small code optimization

r100-stack committed:
Reverted prev commit

r100-stack committed:
Small code optimization

Co-authored-by: Mayank <9084735+mayank99@users.noreply.github.com>

r100-stack committed:
Renamed urgent to pulsing

r100-stack committed:
Changed active prop in JSDocs also to enabled

r100-stack committed:
Changed urgent prop in JSDocs also to pulsing

rohan-kadkol commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,78 @@
+   * Set this to true for important notifications
+   * @default false
+   */
+  urgent?: boolean;
```
Changed `urgent` to `pulsing`. I added a warning to the JSDocs too.

To confirm is there a better way to add the warning? Maybe something like an `@abc` tag that we can use for a warning, or use something similar?

![image](https://user-images.githubusercontent.com/45748283/191336292-dc06e7de-7624-4434-9a14-43eef1d905be.png)

mayank99 replied:
There is nothing in jsdoc, it just comes down to better wording.

r100-stack committed:
Changed type prop in JSDocs also to status

r100-stack committed:
Changed child from div to span in test

r100-stack committed:
Added NotificationMarker to Header's Full story

r100-stack committed:
Added NotificationMarker to SideNavigation's Basic story

rohan-kadkol commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,125 @@
+  active: {
+    control: false,
+  },
+};
```
I added uses of `<NotificationMarker>` in `Header`'s Full story ([commit](https://github.com/iTwin/iTwinUI-react/pull/829/commits/3b029f0c1df338e65334384f13d79c8c7168d3b7)) and `SideNavigations`'s Basic story ([commit](https://github.com/iTwin/iTwinUI-react/pull/829/commits/bee281d467f8697e832763e203652837999daa8e)).

I did not add any props to modify the `NotificationMarker`s since it might be confusing whether `status` and `pulsing` are parameters of `Header`/`SideNavigation` or `NotificationMarker`.

To confirm, is the way I have added it alright?

elephantcatdog replied:
Don't forget to update the story test images for Header and SideNavigation.

rohan-kadkol replied:
**[DONE]** Thanks, approved Header and SideNavigation images

elephantcatdog commented on apps/storybook/src/Header.stories.tsx:
```diff
@@ -165,7 +166,9 @@ export const Full: Story<HeaderProps> = (args) => {
-          <SvgNotification />
+          <NotificationMarker status='positive' pulsing={true}>
+            <SvgNotification />
+          </NotificationMarker>
```
The test image that goes along with this story should be updated.

rohan-kadkol replied:
**[DONE]**

elephantcatdog commented on apps/storybook/src/SideNavigation.stories.tsx:
```diff
@@ -50,18 +51,39 @@ export const Basic: Story<SideNavigationProps> = (args) => {
+            </NotificationMarker>
+          }
+          key={3}
+        >
```
The test image that goes along with this story needs to be updated.

elephantcatdog reviewed (commented):
Don't forget that the programmatic screenshot needs to be removed.

r100-stack committed:
Approved images for Header and SideNavigation

r100-stack committed:
Changed CustomChild story to Pulsing

r100-stack committed:
Approved test image

r100-stack committed:
Deleted unused story test image

rohan-kadkol commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,77 @@
+} from '@itwin/itwinui-react';
+import SvgNotification from '@itwin/itwinui-icons-react/cjs/icons/Notification';
+
+export default {
```
**[DONE]** Nice idea, thanks

elephantcatdog replied:
Hm 🤔 I just thought about how there could be a problem with visual testing the pulse if one time it's 'on' and another time it's 'off'. Have you noticed a problem with that if you run that test a few times?

rohan-kadkol replied:
I thought of that but then realized that the CSS repo also does a similar form of image testing. The [header](https://github.com/iTwin/iTwinUI/blob/main/packages/itwinui-css/backstop/results/bitmaps_reference/iTwinUI_dark_header_Type_All_-_narrow_screen_0_demo-all_0_.png) and [side-navigation](https://github.com/iTwin/iTwinUI/blob/main/packages/itwinui-css/backstop/results/bitmaps_reference/iTwinUI_dark_side-navigation_State_collapsed_hover_and_focus_0_demo-side-navigation_0_desktop.png) tests also have a NotificationMarker in the screenshot. Since it worked there with Backstop, I thought it might work here too with Cypress.

Additionally, I ran the tests a couple of times and they always passed. Also, there are multiple attempts for each test (around 4), so I think they will pass even if they fail in one of the attempts

@mayank99 What do you think?

elephantcatdog replied:
Good info, thanks!

r100-stack committed:
Approve test images

rohan-kadkol commented:
> Are you planning to do the CSS changes (e.g. `mask`) before using it in React?

As of now, I am working on the masking approach on the side. So I think the masking change would come after this PR. 

After making the masking change in CSS, we might need to change the class names and dom structure slightly in react, which I don't think would be a big PR.

r100-stack committed:
Added Status story and modified Basic and Pulsing

rohan-kadkol commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,134 @@
+  pulsing: true,
+};
+
+export const Status: Story<NotificationMarkerProps> = (args) => {
```
Is this story showing too much information? Is there a better way? Or this is alright?

Basically I intended to display all four statuses for text and icons at once, similar to how we show all things at once for most components on the CSS repo's Github demo website.

elephantcatdog replied:
I like the story for the statuses. I think you don't need to show both text and icon. You could do something like this: 
![image](https://user-images.githubusercontent.com/37936169/191810190-e402e962-fbc3-45d1-b917-2ca4d84613e1.png)
That tells users also what the name of the color is.

rohan-kadkol replied:
Thanks for the idea. I added the status names as the dummy names for the text notification examples for the `Status` story. 

However, I kept the icons examples in that story too since I felt it was worth showing both icons and texts. The user can then decide whether they want to use an icon, text, or something else.

rohan-kadkol commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,134 @@
+      </IconButton>
+
+      <NotificationMarker status='primary' {...args}>
+        <Text style={{ fontWeight: '700' }}>URGENT</Text>
```
How to add bold text using our `<Text>` component? Is this the way?

elephantcatdog replied:
In `ComboBox.stories.tsx` the fontWeight is set to 'bold', but it looks like it's the same thing (ie 'bold' == '700'), so I think it's fine as is. Doesn't seem to be another way to do it.
![image](https://user-images.githubusercontent.com/37936169/208146462-3407da14-7212-4127-adb1-54f876a8a68d.png)

r100-stack committed:
Clearer text of Basic and Status story examples

mayank99 commented:
> As of now, I am working on the masking approach on the side. So I think the masking change would come after this PR.
> 
> After making the masking change in CSS, we might need to change the class names and dom structure slightly in react, which I don't think would be a big PR.

If there is not a lot to do in the CSS PR, I'd suggest doing that first. There is no rush to release NotificationMarker so we can wait a bit longer.

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Removed React.cloneElement in IconButton

veekeys committed:
feat!: Update itwinui-css to 1.0.0.dev.10 (#733)

Includes all breaking changes made in itwinui-css so far.

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Merge branch 'v2' into rohan/notification-marker-1

mayank99 committed:
feat(Alert)!: Remove deprecated `onClick` prop (#851)

mayank99 committed:
feat(Tabs)!: Remove deprecated `HorizontalTab` and `HorizontalTabProps` (#852)

mayank99 committed:
feat(Checkbox, Radio)!: Remove deprecated `checkmarkClassName` and `checkmarkStyle` (#855)

gretanausedaite committed:
feat(Table): Add row loading state (#871)

- Added `isLoading` to the row
- Remove `iui-status` for being redundant (adding status with `data-iui-status`)

adamhock committed:
feat(Table): Row selection count for paginator (#837)

* There is now a counter in the bottom left section of paginated tables that shows how many rows have been selected if multi-selection mode is enabled

r100-stack committed:
feat(Avatar, AvatarGroup): Deprecate `UserIcon`/`UserIconGroup`, replace with `Avatar`/`AvatarGroup` (#902)

* Deprecated and renamed `UserIcon`/`UserIconGroup` to `Avatar`/`AvatarGroup`

gretanausedaite committed:
feat(Tile): Add status and loading state (#872)

Add status, loading and disabled states to Tile.

elephantcatdog committed:
feat(Table): example for initial state (#882)

mayank99 committed:
chore: Create new docs site (#785)

The code lives in a new workspace called website. This is just an initial version. There are missing features, bugs, and the content is placeholder. But most of the code (excluding content) should be good to keep in the final version.

Co-authored-by: Mayank <mayank99@users.noreply.github.com>
Co-authored-by: Greta <greta.nausedaite@bentley.com>

r100-stack committed:
feat(Wizard): Deprecate `Wizard` to use `Stepper`/`WorkflowDiagram` (#905)

* Deprecated `Wizard` to use `Stepper` (`Wizard.type = default | long`) or WorkflowDiagram (`Wizard.type = workflow`)
* Deprecated helpers to `Wizard`:
  * `WizardProps`:  Use `StepperProps` with `Stepper`
  * `WizardType`: Use `StepperType` (`Wizard.type = default | long`)
  * `WizardLocalization`: Use `StepperLocalization`

r100-stack committed:
Bump itwinui-css to 1.0.0-dev.12

r100-stack committed:
Updated NotificationMarker.tsx to work with new css changes

r100-stack committed:
Updated NotificationMarker test

r100-stack committed:
Merge branch 'v2' into rohan/notification-marker-1

r100-stack committed:
Bump itwinui-css to 1.0.0-dev.13

r100-stack committed:
Reverted IconButton to use cloneElement

r100-stack committed:
feat(Tabs): Deprecate `HorizontalTabs`, `VerticalTabs` to use `Tabs` (#908)

* Deprecated `HorizontalTabs`, `VerticalTabs` to directly use `Tabs`
* Deprecated `HorizontalTabsProps`, `VerticalTabsProps` to use `TabsProps` with `Tabs`
* Added `TabsOrientationProps` to control `type` depending on `orientation`
  * `TabsProps` includes `TabsOrientationProps`

dependabot[bot] committed:
chore(deps): bump loader-utils from 1.4.0 to 1.4.1 (#918)

gretanausedaite committed:
fix(Header): Change `iui-slim` to `data-iui-size: slim` (#919)

mayank99 committed:
chore: Bump vulnerable node-fetch (#921)

mayank99 committed:
docs: Rename Wizard to Stepper (#922)

Needed for build to pass. Sandbox will be fixed in future.

r100-stack committed:
fix(Button): Remove `cloneElement()` (#895)

* Removed `React.cloneElement()` from `Button` (`startIcon`, `endIcon`) and `IconButton`.
* Instead uses a `<span className="iui-button-icon" aria-hidden>` wrapper.

r100-stack committed:
Reverted itwinui-css to use 1.0.0-dev.12

r100-stack committed:
Merge branch 'v2' into rohan/notification-marker

r100-stack committed:
Changed logic for enabled=false. Even when notification is disabled, still have a wrapper span. I feel this will be needed for avoiding formatting issues. E.g. maybe some selector that counts on position might become wrong when notification span is there when enabled but not there when disabled

r100-stack committed:
Pasted files from main

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
GitLens reverted images differences

r100-stack committed:
Added most changes from before

r100-stack committed:
Re-added old NotificationMarker stories

r100-stack committed:
Added WithNotification story to Button

r100-stack committed:
More apt icon for WithNotification Button story

r100-stack committed:
Added 'white' as valid status

r100-stack committed:
Improved JSDocs formatting

r100-stack committed:
Added four buttons to show NotificationMarker.style = white

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Fixed failing unit test

r100-stack committed:
Approved the failing NotificationMarker storybook test images. Need to fix later

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

rohan-kadkol commented on apps/storybook/src/NotificationMarker.test.ts:
```diff
@@ -0,0 +1,16 @@
+    it(testName, function () {
+      const id = Cypress.storyId(storyPath, testName);
+      cy.visit('iframe', { qs: { id } });
+      cy.compareSnapshot(testName);
```
I could not fix the NotificationMarker test images. In the `NotificationMarker -> Status` story for example, the notification marker circle appears on the top right when viewing in Storybook.

![image](https://user-images.githubusercontent.com/45748283/207117469-16148deb-918f-4965-8cf9-a2b0ae7f9035.png)

But the test images show the circle on the top center for the `IconButton`.

![image](https://user-images.githubusercontent.com/45748283/207117359-efa70977-a071-44e9-b500-bb67546531ca.png)

<!--![image](https://user-images.githubusercontent.com/45748283/207117291-f5393aac-349e-4334-9060-8003f790cd8e.png)-->

I tried using @mayank99 's suggestion to use `cy.wait()` but that didn't work. Does anyone have any ideas on why this is happening and how to fix this?

mayank99 replied:
After spending some time on this, I was able to narrow down the issue to `IconButton` but couldn't really figure out the exact reason it gets misaligned like that. Could be a bug in Chrome 94 (which is what Cypress 9.6.0 comes with).

I was able to work around it by removing `IconButton` and instead adding the fill/size attributes on the svg (which will not be necessary after https://github.com/iTwin/iTwinUI/issues/851 and/or https://github.com/iTwin/iTwinUI-icons/issues/55). I've pasted all my local changes below, feel free to copy/change as you see fit. Hopefully this unblocks you for now.

<details>

<summary>NotificationMarker.stories.tsx</summary>

```tsx
/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
import { Story, Meta } from '@storybook/react';
import React from 'react';
import {
  Text,
  NotificationMarker,
  NotificationMarkerProps,
} from '@itwin/itwinui-react';
import { SvgNotification } from '@itwin/itwinui-icons-react';

export default {
  component: NotificationMarker,
  args: {
    enabled: true,
    pulsing: false,
  },
  argTypes: {
    className: { control: { disable: true } },
    style: { control: { disable: true } },
    status: {
      control: 'radio',
      options: ['primary', 'positive', 'warning', 'negative'],
    },
  },
  title: 'Core/NotificationMarker',
  decorators: [
    (Story) => (
      <div
        style={{
          display: 'grid',
          placeItems: 'start',
          gap: '20px',
        }}
      >
        <Story />
      </div>
    ),
  ],
} as Meta<NotificationMarkerProps>;

export const Basic: Story<NotificationMarkerProps> = (args) => {
  return (
    <>
      <NotificationMarker {...args}>
        <SvgNotification width={16} height={16} fill='currentColor' />
      </NotificationMarker>
      <NotificationMarker {...args}>Normal</NotificationMarker>
    </>
  );
};

Basic.args = {
  status: 'primary',
  children: <SvgNotification />,
};

Basic.argTypes = {
  children: {
    control: false,
  },
};

export const Pulsing: Story<NotificationMarkerProps> = (args) => {
  return (
    <>
      <NotificationMarker status='primary' {...args}>
        <SvgNotification width={16} height={16} fill='currentColor' />
      </NotificationMarker>

      <NotificationMarker status='primary' {...args}>
        <Text style={{ fontWeight: '700' }}>URGENT</Text>
      </NotificationMarker>
    </>
  );
};

Pulsing.args = {
  status: 'negative',
  pulsing: true,
};

export const Status: Story<NotificationMarkerProps> = (args) => {
  return (
    <>
      <NotificationMarker status='primary' {...args}>
        <SvgNotification width={16} height={16} fill='currentColor' />
      </NotificationMarker>
      <NotificationMarker status='positive' {...args}>
        <SvgNotification width={16} height={16} fill='currentColor' />
      </NotificationMarker>
      <NotificationMarker status='warning' {...args}>
        <SvgNotification width={16} height={16} fill='currentColor' />
      </NotificationMarker>
      <NotificationMarker status='negative' {...args}>
        <SvgNotification width={16} height={16} fill='currentColor' />
      </NotificationMarker>
      <NotificationMarker status='primary' {...args}>
        Primary
      </NotificationMarker>
      <NotificationMarker status='positive' {...args}>
        Positive
      </NotificationMarker>
      <NotificationMarker status='warning' {...args}>
        Warning
      </NotificationMarker>
      <NotificationMarker status='negative' {...args}>
        Negative
      </NotificationMarker>
    </>
  );
};

Status.argTypes = {
  status: {
    control: false,
  },
};
```
</details>

rohan-kadkol replied:
Thanks. I ended up adding a story called `NotificationMarker -> Basic` that directly uses an SVG (e.g. `<SvgNotification ... />`).

The other three stories in `NotificationMarker` use `IconButton`/`Button` and so the visual tests still give the above-mentioned bug. Thus, these three stories are left out of the visual tests.

The visual test for `Button -> WithNotification` is excluded from visual tests for the same reason.

rohan-kadkol replied:
**[DONE]** Added comments in `NotificationMarker.test.ts` and `Button.test.ts` explaining why certain tests are left out

r100-stack committed:
Removed IconButton in NotificationMarker stories to avoid image test errors

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Removed NotificationMarker from Header and SideNav since they require IconButton but IconsButton causes  NotificationMarker misalignment in visual testing

r100-stack committed:
Updated yarn.lock

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Merge branch 'rohan/notification-marker' of https://github.com/iTwin/iTwinUI-react into rohan/notification-marker

r100-stack committed:
Reverted back to having IconButton in all NotificationMarker stories except Basic

r100-stack committed:
Added "white" as available option for status

r100-stack committed:
Removed unnecessary "children" story control

r100-stack committed:
Made Button in WithButtons story to have no startIcon and notification marker on the child instead of the startIcon

r100-stack committed:
Added onClick that calls action('clicked') to all buttons

r100-stack committed:
NotificationMarker only tests for Basic story that doesn't contain IconButton or Button

r100-stack committed:
Deleted old NotificationMarker test images. Added new NotificationMarker Basic story test image (didn't change so no real addition)

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,85 @@
+   * </NotificationMarker>
+   */
+  enabled?: boolean;
+} & Omit<CommonProps, 'title'>;
```
this should take all props 
```suggestion
} & React.ComponentProps<'span'>;
```

rohan-kadkol replied:
**[DONE]**

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,85 @@
+ * @example
+ * <NotificationMarker status='positive' pulsing={true}>Live</NotificationMarker>
+ */
+export const NotificationMarker = (props: NotificationMarkerProps) => {
```
forward ref might help here if the user wants to access the dom element

```suggestion
export const NotificationMarker = React.forwardRef(props: NotificationMarkerProps, ref: React.Ref<HTMLSpanElement>) => {
```

(the suggestion will need formatting and closing brackets)

rohan-kadkol replied:
Added `React.forwardRef`

r100-stack committed:
Added comments with reasons for selectively skipping visual tests

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,85 @@
+  } = props;
+  useTheme();
+
+  const notificationAttributes = enabled
+    ? {
+        'data-iui-variant': status,
+        'data-iui-urgent': pulsing,
+      }
+    : {};
+
+  return (
+    <span
+      className={cx({ 'iui-notification-marker': enabled }, className)}
+      {...notificationAttributes}
```
why not apply the data attributes directly?
```suggestion
  return (
    <span
      className={cx({ 'iui-notification-marker': enabled }, className)}
      data-iui-variant={enabled && status}
      data-iui-urgent={enabled && pulsing)
```

rohan-kadkol replied:
I wanted to not include the data attributes when the `NotificationMarker` is disabled. To keep the same functionality and also reduce the lines of code, I set the data attributes to `null` inline when `enabled = false`

```tsx
data-iui-variant={enabled ? status : null}
data-iui-urgent={enabled ? pulsing : null}
```

mayank99 commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,147 @@
+export const WithButtons: Story<NotificationMarkerProps> = (args) => {
+  return (
+    <>
+      <IconButton styleType='borderless' onClick={action('clicked')}>
```
no need for click action imo. we want to show the notification marker, so the less code in the surrounding elements, the better

rohan-kadkol replied:
Removed

r100-stack committed:
Added React.forwardRef to NotificationMarker. Corrected props type. Improved code.

r100-stack committed:
Removed all onClick from NotificationMarker stories

r100-stack committed:
Exclude data attributes when enabled = false

mayank99 commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,84 @@
+ * @example
+ * <NotificationMarker status='positive' pulsing={true}>Live</NotificationMarker>
+ */
+// export const NotificationMarker = (props: NotificationMarkerProps) => {
```
leftover

rohan-kadkol replied:
Thanks, removed

r100-stack committed:
Removed unnecessary comment and leftover code

rohan-kadkol commented on packages/iTwinUI-react/src/core/NotificationMarker/NotificationMarker.tsx:
```diff
@@ -0,0 +1,84 @@
+      <span
+        ref={ref}
+        className={cx({ 'iui-notification-marker': enabled }, className)}
+        // Exclude data attributes when enabled = false
```
Removed

r100-stack committed:
Improved JSDocs formatting

elephantcatdog commented on apps/storybook/src/Button.stories.tsx:
```diff
@@ -87,6 +88,69 @@ WithIcon.args = {
+  children: 'Inbox',
+};
+
+WithIcon.argTypes = {
```
I think this should be "WithNotification.argTypes"?

rohan-kadkol replied:
**[DONE]**

r100-stack committed:
Fixed incorrect story name

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

elephantcatdog commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,146 @@
+        Negative
+      </NotificationMarker>
+    </>
+  );
```
Maybe add an example for 'white' status in here? It'd probably be helpful visually to go along with explanation 
![image](https://user-images.githubusercontent.com/37936169/208144261-40c4614f-8176-4930-bfc2-9993d7f6e441.png)

elephantcatdog replied:
I found an example in `Button.stories.tsx` that has 'white' status. Even though it's an example there, I think it'd still be relevant here if you want to just copy it over or do something similar.

rohan-kadkol replied:
I added examples for `styleType = white` in `NotificationMarker -> Status` story. I also changed the grid layout to make it more intuitive.

<img width="300px" src="https://user-images.githubusercontent.com/45748283/208152725-4c7d19c4-96d0-4c49-b3a0-a09d270b8104.png" />

elephantcatdog replied:
Looks good; I like it!

elephantcatdog commented on apps/storybook/src/NotificationMarker.stories.tsx:
```diff
@@ -0,0 +1,146 @@
+import {
+  Text,
+  NotificationMarker,
+  NotificationMarkerProps,
```
Want to confirm that this issue is just on my end: 
![image](https://user-images.githubusercontent.com/37936169/208146137-7c313e20-efa2-4b94-b904-e999ad0c46b3.png)

rohan-kadkol replied:
This is not happening to me, but it might be alright. I too get similar errors sometimes but it occasionally goes away after restarting VSCode or re-running `yarn install`. Mayank also suggested to try running the `TypeScript: Restart TS Server` VSCode command to see if it helps, since Storybook is usually unreliable.

But for now, I think this error can be ignored.

r100-stack committed:
Added 'white' examples to Status story. In Status story, each status is a row

r100-stack committed:
Changed column and row gap

r100-stack committed:
Merge branch 'main' into rohan/notification-marker

r100-stack committed:
Merge branch 'rohan/notification-marker' of https://github.com/iTwin/iTwinUI-react into rohan/notification-marker
